include/time_series/internal/base.hxx

```diff
@@ -323,3 +323,67 @@ void TimeSeriesBase<P, T>::monitor_signal()
     // terminating.
     local_condition_ptr->notify_all();
 }
+
+template <typename P, typename T>
+std::vector<std::tuple<T, Index, Timestamp>> TimeSeriesBase<P, T>::snapshot()
+    const
+{
+    std::vector<std::tuple<T, Index, Timestamp>> snap;
+    int max_length = this->max_length();
+
+    Lock<P> lock(*this->mutex_ptr_);
+    read_indexes();
+
+    T element;
+    Timestamp timestamp;
+
+    // the time series underlying data structure is not yet full
+    if (newest_timeindex_ < max_length)
+    {
+        // these indexes have been set with an element
+        for (long int i = 0; i <= newest_timeindex_; i++)
+        {
+            this->history_elements_ptr_->get(i, element);
+            this->history_timestamps_ptr_->get(i, timestamp);
+            snap.push_back(std::make_tuple(element, i, timestamp));
+        }
+        if (newest_timeindex_ == max_length - 1)
+        {
+            snap.resize(max_length);
+            return snap;
+        }
+        // these indexes have not yet been set with an element
+        // timestamp at value -1 indicates this to the end-user.
+        for (long int i = newest_timeindex_ + 1; i < max_length; i++)
+        {
+            snap.push_back(std::make_tuple(T(), -1, -1));
+        }
+        snap.resize(max_length);
+        return snap;
+    }
```

Comment:
I think using an `else`-block for the code below and only having a single `return` at the end of the function would make it a bit easier to follow the flow of the code.  It might just be a matter of taste, though.